lib/services/role.go

```diff
@@ -1887,69 +1600,174 @@ func (m *DatabaseNameMatcher) String() string {
 	return fmt.Sprintf("DatabaseNameMatcher(Name=%v)", m.Name)
 }
 
-// CheckAccessToDatabase checks if this role set has access to a particular database.
-//
-// The checker always checks the server namespace, other matchers are supplied
-// by the caller.
-func (set RoleSet) CheckAccessToDatabase(server types.Database, mfa AccessMFAParams, matchers ...RoleMatcher) error {
+type loginMatcher struct {
+	login string
+}
+
+// NewLoginMatcher creates a RoleMatcher that checks whether the role's logins
+// match the specified condition.
+func NewLoginMatcher(login string) RoleMatcher {
+	return &loginMatcher{login: login}
+}
+
+func (l *loginMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
+	logins := role.GetLogins(typ)
+	for _, login := range logins {
+		if l.login == login {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// AccessCheckable is the subset of types.Resource required for the RBAC checks.
+type AccessCheckable interface {
```

Comment:
Why not just use `types.Resource`? It's our base interface for all resources.

Reply:
Because:

1) it's huge - there's way more methods in there than we actually need, making it harder to mock out and test
2) `types.KubernetesCluster` is not a `types.Resource`, so the wrapper necessary to make that work would require us to stub out even more methods we never plan to call.
2) We actually need `types.ResourceWithLabels`, not `types.Resource`, but unfortunately that interface also embeds `Origin() string` which not all resources implement, requiring further updates to all resources.

Reply:
For mocking you can just embed `types.Resource` into your test struct and only implement the methods you need and with my suggestion below KubeCluster will also be a resource. But it's no big deal, specialized interfaces are fine.